src/raster/rasterband.rs

```diff
@@ -636,29 +600,32 @@ impl<'a> RasterBand<'a> {
     ///         &options,
     ///     )?;
     /// let mut band1 = dataset.rasterband(1)?;
-    /// let arr = Array2::from_shape_fn((16, 16), |(y, x)| y as u16 * 16 + x as u16);
-    /// band1.write_block((0, 0), arr)?;
+    /// let arr = Buffer::new((16, 16), (0..16*16).collect());
+    /// band1.write_block((0, 0), &arr.into())?;
     /// # Ok(())
     /// # }
     /// ```
     pub fn write_block<T: Copy + GdalType>(
         &mut self,
         block_index: (usize, usize),
-        block: Array2<T>,
+        block: &Buffer<T>,
     ) -> Result<()> {
         if T::gdal_ordinal() != self.band_type() as u32 {
             return Err(GdalError::BadArgument(
                 "array type must match band data type".to_string(),
             ));
         }
 
-        let mut data = block.into_raw_vec();
         let rv = unsafe {
             gdal_sys::GDALWriteBlock(
                 self.c_rasterband,
                 block_index.0 as c_int,
                 block_index.1 as c_int,
-                data.as_mut_ptr() as GDALRasterBandH,
+                // This parameter is marked as `* mut c_void` because the C/C++ API for some reason
+                // doesn't mark it as `const void *`. From code inspection starting at the link
+                // below, it appears to be a read-only array.
+                // https://github.com/OSGeo/gdal/blob/b5d004fb9e3fb576b3ccf5f9740531b0bfa87ef4/gcore/gdalrasterband.cpp#L688
+                block.data.as_ptr() as *mut c_void,
```

Comment:
@lnicola I have a pending question to the GDAL devs on this:

https://lists.osgeo.org/pipermail/gdal-dev/2023-December/058130.html

I would have expected a "write" method to only need a shared reference/pointer to the data.

Reply:
@lnicola Based on this response, do you think the above is OK? Seems like our only options are:

1. Make `block` `&mut`
2. Create a copy of `block` before passing the data along
3. Accept the ambiguity and hope no-one runs into problems :/ 

https://lists.osgeo.org/pipermail/gdal-dev/2023-December/058134.html

Reply:
`&mut` with a note in the comments seems perfectly reasonable.